x/delayedack/ibc_middleware.go

```diff
@@ -293,6 +319,74 @@ func (im IBCMiddleware) OnTimeoutPacket(
 	return nil
 }
 
+// createDemandOrderFromIBCPacket creates a demand order from an IBC packet.
+// It validates the fungible token packet data, extracts the fee from the memo,
+// calculates the demand order price, and creates a new demand order.
+// It returns the created demand order or an error if there is any.
+func (im IBCMiddleware) createDemandOrderFromIBCPacket(fungibleTokenPacketData transfertypes.FungibleTokenPacketData,
+	rollappPacket *types.RollappPacket, rollappPacketStoreKey string, memoObj map[string]interface{}) (*eibctypes.DemandOrder, error) {
+	// Validate the fungible token packet data as we're going to use it to create the demand order
+	if err := fungibleTokenPacketData.ValidateBasic(); err != nil {
+		return nil, err
+	}
+	// Verify the original recipient is not a blocked sender otherwise could potentially use eibc to bypass it
+	if im.keeper.BlockedAddr(fungibleTokenPacketData.Receiver) {
+		return nil, fmt.Errorf("%s is not allowed to receive funds", fungibleTokenPacketData.Receiver)
+	}
+	// Get the fee from the memo
+	fee := memoObj[eibcMemoObjectName].(map[string]interface{})[eibcMemoFieldFee].(string)
```

Comment:
the memo should be parsed to struct and validated before usage
(https://github.com/cosmos/ibc-apps/blob/83f866e7f1d2577477ce3ad38c67edb581ce29b4/middleware/packet-forward-middleware/packetforward/ibc_middleware.go#L191)